File: go-service/internal/domain/product_test.go
package domain

import "testing"

func TestProductValidate(t *testing.T) {
	t.Run("valid product", func(t *testing.T) {
		p := &Product{Name: "Keyboard", Price: 10}
		if err := p.Validate(); err != nil {
			t.Fatalf("expected no error, got %v", err)
		}
		if len(p.Tags) != 0 {
			t.Fatalf("expected tags to be initialized")
		}
	})

	t.Run("missing name", func(t *testing.T) {
		p := &Product{Price: 10}
		err := p.Validate()
		if err == nil || err.Code != ErrorCodeValidation {
			t.Fatalf("expected validation error for missing name")
		}
	})

	t.Run("invalid price", func(t *testing.T) {
		p := &Product{Name: "Keyboard", Price: 0}
		err := p.Validate()
		if err == nil || err.Code != ErrorCodeValidation {
			t.Fatalf("expected validation error for invalid price")
		}
	})
}

File: node-service/tests/domain/product.test.js
const test = require('node:test');
const assert = require('node:assert');

const Product = require('../../internal/domain/product');
const DomainError = require('../../internal/domain/errors');

test('product validation succeeds', () => {
  const product = new Product({ name: 'Keyboard', price: 100 });
  assert.doesNotThrow(() => product.validate());
  assert.deepStrictEqual(product.tags, []);
});

test('product validation fails for invalid input', () => {
  const product = new Product({ name: '', price: -10 });
  assert.throws(() => product.validate(), (err) => {
    assert.ok(err instanceof DomainError);
    assert.strictEqual(err.code, 'validation');
    return true;
  });
});
